### PR TITLE
Harnesses verifying slice types for add, sub and offset

### PR DESCRIPTION
**Pull request by szlee118**
Towards https://github.com/model-checking/verify-rust-std/issues/76

### Contracts Added:

- Added contracts for <*mut T>::add, <*mut T>::sub, and <*mut T>::offset operations for raw pointers to slices.
### Proofs Implemented:

- Added proofs for the above contracts, verifying safety for the following slice pointee types using arrays:
 - All integer types (i8, i16, i32, etc.).
 - Tuples (composite types) (e.g., (i8, i8), (i32, f64, bool)).
 - Unit type (()).
 
### Macro Definitions:
- Introduced macros to simplify and automate harness generation:
 - One macro for add and sub operations on slices.
 - Another macro for offset operations on slices.
 
### Array-Based Implementation for Slice Verification:
- Arrays are used as a proxy for slices (as slices do not have a known length at compile time) to enable proper bounds checking and assumptions during verification.

### This should be merged after [another PR](https://github.com/model-checking/verify-rust-std/pull/113/files) for dependency issues.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by szlee118** [comment]
@feliperodri FYI

**Comment by tautschnig** [comment]
@szlee118 #113 has been merged, but now conflicts need to be resolved on this one.

**Comment by szlee118** [comment]
@tautschnig @feliperodri conflict resolved.

File: library/core/src/ptr/mut_ptr.rs
Comment (feliperodri) [inline library/core/src/ptr/mut_ptr.rs:2531]: If we add this line, the comment below should be removed and the issue solved.

File: library/core/src/ptr/mut_ptr.rs
Comment (feliperodri) [inline library/core/src/ptr/mut_ptr.rs:2539]: duplicated line

File: library/core/src/ptr/mut_ptr.rs
Comment (feliperodri) [inline library/core/src/ptr/mut_ptr.rs:2572]: duplicated line

**Review by feliperodri — COMMENTED** [review]
It seems there are still some conflicts. Could you check if the conflicts were successfully solved?

File: library/core/src/ptr/mut_ptr.rs
Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs]: Line removed in newest commit.

File: library/core/src/ptr/mut_ptr.rs
Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs:2570]: Line removed in newest commit.

File: library/core/src/ptr/mut_ptr.rs
Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs]: Line removed in newest commit.

File: library/core/src/ptr/mut_ptr.rs
Comment (feliperodri) [inline library/core/src/ptr/mut_ptr.rs:2459]: It's missing the key part of the explanation: why? why 1 keeps `ptr_with_offset` for these types?

File: library/core/src/ptr/mut_ptr.rs
Comment (feliperodri) [inline library/core/src/ptr/mut_ptr.rs:2372]: Isn't 5 too low? How is the verification performance numbers?

Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs:2372]: Explained in another [PR's comment](https://github.com/model-checking/verify-rust-std/pull/166#discussion_r1861482793) that 5 is enough for this case.
This `performance` statement is outdated, and has been removed by newest commit.

File: library/core/src/ptr/mut_ptr.rs
Comment (xsxszab) [reply library/core/src/ptr/mut_ptr.rs:2459]: This is a legacy implementation for the proof of contract. In our latest implementation (detailed in PR #166), it has been updated to use the pointer generator.

**Comment by szlee118** [comment]
> It seems there are still some conflicts. Could you check if the conflicts were successfully solved?

It has been solved successfully.

**Review by feliperodri — APPROVED** [review]

File: library/core/src/ptr/mut_ptr.rs
Comment (tautschnig) [inline library/core/src/ptr/mut_ptr.rs:407]: Why are you introducing fresh blank lines (here and below)?

File: library/core/src/ptr/mut_ptr.rs
Comment (tautschnig) [inline library/core/src/ptr/mut_ptr.rs:1039]: No need for adding whitespace.

File: library/core/src/ptr/mut_ptr.rs
Comment (tautschnig) [reply library/core/src/ptr/mut_ptr.rs:2537]: Still adds whitespace at the end of the line.

File: library/core/src/ptr/mut_ptr.rs
Comment (tautschnig) [inline library/core/src/ptr/mut_ptr.rs:2505]: Adding whitespace.

File: library/core/src/ptr/mut_ptr.rs
Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs]: Removed in newest commit.

File: library/core/src/ptr/mut_ptr.rs
Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs]: Removed in newest commit.

File: library/core/src/ptr/mut_ptr.rs
Comment (tautschnig) [inline library/core/src/ptr/mut_ptr.rs:2366]: Could you please add a further comment saying _why_ you have chosen this particular value?

Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs]: Comment added.

**Comment by feliperodri** [comment]
@tautschnig @carolynzech could you do another round of review here? We need one more approval before merging it.

**Pull request merged**
